{"metadata":{"kernelspec":{"language":"python","display_name":"Python 3","name":"python3"},"language_info":{"name":"python","version":"3.7.10","mimetype":"text/x-python","codemirror_mode":{"name":"ipython","version":3},"pygments_lexer":"ipython3","nbconvert_exporter":"python","file_extension":".py"}},"nbformat_minor":4,"nbformat":4,"cells":[{"cell_type":"code","source":"import numpy as np # linear algebra\nimport pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)\nimport re\nimport pymc3 as pm\nimport theano.tensor as tt\nimport string\nimport nltk\nimport matplotlib.pyplot as plt\nimport matplotlib.cm as cm\nimport seaborn as sns\nsns.set_style('darkgrid')\nimport plotly.express as ex\nimport plotly.graph_objs as go\nimport plotly.offline as pyo\nfrom plotly.subplots import make_subplots\npyo.init_notebook_mode()\nfrom sklearn.decomposition import TruncatedSVD,PCA\nfrom sklearn.cluster import KMeans\nimport matplotlib.gridspec as gridspec\nimport random\nfrom tqdm.notebook import tqdm\nimport gc\nfrom scipy.stats.mstats import mquantiles\n%pip install joypy\nfrom joypy import joyplot\nplt.rc('figure',figsize=(17,10))\nsns.set_context('paper',font_scale=2)","metadata":{"execution":{"iopub.status.busy":"2021-11-10T05:15:48.759253Z","iopub.execute_input":"2021-11-10T05:15:48.759569Z","iopub.status.idle":"2021-11-10T05:15:57.848807Z","shell.execute_reply.started":"2021-11-10T05:15:48.759532Z","shell.execute_reply":"2021-11-10T05:15:57.848042Z"},"trusted":true},"execution_count":3,"outputs":[]},{"cell_type":"code","source":"s_data = pd.read_csv('/kaggle/input/cpg-values-of-smoking-and-non-smoking-patients/Smoker_Epigenetic_df.csv')\ns_data.Gender = s_data.Gender.str.lower()\ns_data.drop(columns=['GSM'],inplace=True)\ns_data.head(3)","metadata":{"execution":{"iopub.status.busy":"2021-11-10T05:26:00.832602Z","iopub.execute_input":"2021-11-10T05:26:00.832879Z","iopub.status.idle":"2021-11-10T05:26:00.889486Z","shell.execute_reply.started":"2021-11-10T05:26:00.832848Z","shell.execute_reply":"2021-11-10T05:26:00.888829Z"},"trusted":true},"execution_count":5,"outputs":[]},{"cell_type":"code","source":"plt.title('Number of Missing Values at Each Feature')\nsns.heatmap(s_data.isna().sum().to_frame(),cmap='coolwarm',linewidth=2,annot=True)\ns_data.dropna(inplace=True)\nplt.show()","metadata":{"execution":{"iopub.status.busy":"2021-11-10T05:26:31.686680Z","iopub.execute_input":"2021-11-10T05:26:31.687078Z","iopub.status.idle":"2021-11-10T05:26:32.251423Z","shell.execute_reply.started":"2021-11-10T05:26:31.687033Z","shell.execute_reply":"2021-11-10T05:26:32.250665Z"},"trusted":true},"execution_count":6,"outputs":[]},{"cell_type":"markdown","source":"* there are 62 samples in our dataset the are missing all their genetic features; as we are interested in exploring genetic attributes, we will drop the missing samples in this stage, but it may be interesting to try and treat those samples as a test set for a predictive model and try to replace the missing value with the predictive model.","metadata":{}},{"cell_type":"code","source":"\nfig = plt.figure()\n\nax1 = plt.subplot(221)\nax1.set_title('Distribution of Genders')\nsns.countplot(x=s_data['Gender'],ax=ax1,palette=['tab:pink','tab:blue'])\nax2 = plt.subplot(222)\nax2.set_title('Distribution of Smoking Labels')\nsns.countplot(x=s_data['Smoking Status'],ax=ax2)\nax3 = plt.subplot(212)\nax3.set_title('Distribution of Sample Ages')\nsns.histplot(data=s_data['Age'],ax=ax3,kde=True)\n\n\nplt.tight_layout()\nplt.show()","metadata":{"execution":{"iopub.status.busy":"2021-11-10T05:26:51.972350Z","iopub.execute_input":"2021-11-10T05:26:51.972620Z","iopub.status.idle":"2021-11-10T05:26:52.549925Z","shell.execute_reply.started":"2021-11-10T05:26:51.972592Z","shell.execute_reply":"2021-11-10T05:26:52.549211Z"},"trusted":true},"execution_count":7,"outputs":[]},{"cell_type":"markdown","source":"**Observation:** depending on our questions of interest, we will have to take in mind that our dataset is imbalanced both gender-wise and smoking status wise as well as our dataset age distribution appearing to be negatively skewed; such skewness reduces the confidence of our inference and models among younger patients as the average age is centered around 55.","metadata":{}},{"cell_type":"code","source":"joyplot(\n    data=s_data[list(s_data.columns[4:])], \n    figsize=(15, 12),\n    alpha=0.85\n    ,title='Difference in Probe Methylation Distribution Across Given Cites'\n\n)\nplt.show()","metadata":{"execution":{"iopub.status.busy":"2021-11-10T05:27:07.247311Z","iopub.execute_input":"2021-11-10T05:27:07.247570Z","iopub.status.idle":"2021-11-10T05:27:08.892301Z","shell.execute_reply.started":"2021-11-10T05:27:07.247540Z","shell.execute_reply":"2021-11-10T05:27:08.888314Z"},"trusted":true},"execution_count":8,"outputs":[]},{"cell_type":"markdown","source":"**Observation:** looking at the distribution of each individual probe we can see that many of our probes follow a bimodal distribution this can point out to 2 distinct underlying groups in our data. We have 2 known to us groups: smokers and non-smokers and females/males, but it is not exclusive to those groups. We may uncover an underlying group originating on the ages of our patients, for example","metadata":{}},{"cell_type":"code","source":"\njoyplot(\n    data=s_data[list(s_data.columns[4:])+['Smoking Status']], \n    figsize=(13, 8),\n    by='Smoking Status',\n    alpha=0.85\n    ,title='Difference in Probe Methylation Between Smoking Status'\n\n)\nplt.show()","metadata":{"execution":{"iopub.status.busy":"2021-11-10T05:27:28.852831Z","iopub.execute_input":"2021-11-10T05:27:28.853498Z","iopub.status.idle":"2021-11-10T05:27:29.721158Z","shell.execute_reply.started":"2021-11-10T05:27:28.853464Z","shell.execute_reply":"2021-11-10T05:27:29.720489Z"},"trusted":true},"execution_count":9,"outputs":[]},{"cell_type":"markdown","source":"**Observation:** when looking at the difference in the distribution of methylation level in our probes, we see no significant visible difference between smokers and non-smokers.","metadata":{}},{"cell_type":"code","source":"joyplot(\n    data=s_data[list(s_data.columns[4:])+['Gender']], \n    figsize=(13, 8),\n    by='Gender',\n    alpha=0.85,title='Difference in Probe Methylation Between Genders'\n)\nplt.show()","metadata":{"execution":{"iopub.status.busy":"2021-11-10T05:27:41.586351Z","iopub.execute_input":"2021-11-10T05:27:41.586617Z","iopub.status.idle":"2021-11-10T05:27:42.478304Z","shell.execute_reply.started":"2021-11-10T05:27:41.586587Z","shell.execute_reply":"2021-11-10T05:27:42.477590Z"},"trusted":true},"execution_count":10,"outputs":[]},{"cell_type":"markdown","source":"**Observation:** We can see that when looking at the difference in distribution based on gender, there is an amazingly significant difference between the two genders, the genetic explanations behind this difference are unclear to me as we do not have the associated genes to our probes in our dataset.","metadata":{}},{"cell_type":"code","source":"#Encoding Categorical Features\ns_data.Gender = s_data.Gender.astype('category').cat.codes\ns_data['Smoking Status'] = s_data['Smoking Status'].astype('category').cat.codes","metadata":{"execution":{"iopub.status.busy":"2021-11-10T05:28:53.941507Z","iopub.execute_input":"2021-11-10T05:28:53.941753Z","iopub.status.idle":"2021-11-10T05:28:53.948426Z","shell.execute_reply.started":"2021-11-10T05:28:53.941726Z","shell.execute_reply":"2021-11-10T05:28:53.947729Z"},"trusted":true},"execution_count":12,"outputs":[]},{"cell_type":"code","source":"cx = sns.clustermap(np.round(s_data.corr(),2),linewidth=0.8,cmap='vlag',figsize=(15,15),annot=True,annot_kws=dict(fontsize=11))\ncx.ax_row_dendrogram.set_visible(False)\ncx.ax_col_dendrogram.set_visible(False)\ncx.fig.suptitle('Pearson Correlation Between Features') \ncx.fig.tight_layout()\nplt.show()","metadata":{"execution":{"iopub.status.busy":"2021-11-10T05:29:06.787404Z","iopub.execute_input":"2021-11-10T05:29:06.787851Z","iopub.status.idle":"2021-11-10T05:29:10.143177Z","shell.execute_reply.started":"2021-11-10T05:29:06.787809Z","shell.execute_reply":"2021-11-10T05:29:10.139850Z"},"trusted":true},"execution_count":13,"outputs":[]},{"cell_type":"markdown","source":"**Observation:** looking at the clustered Pearson correlations, we see many features correlated to each other, especially our probes. Such an observation leads us to believe that we have multicollinearity in our data and that we most likely cannot assume independence between all the probes we are working with. Next, we will try and reduce the dimensionality of our data and try to confirm our hypothesis. If the same data in a reduced dimension will have a high EVR, we will continue our analysis with the more appropriate reduced dimension.","metadata":{}},{"cell_type":"code","source":"pca = PCA(2)\ntransformed = pca.fit_transform(s_data.iloc[:,4:])\nt_df = pd.DataFrame(transformed,columns=['pc1','pc2'])\nt_df['Gender'] = s_data.Gender\nt_df['Age'] = s_data.Age\nt_df['Smoking Status'] = s_data['Smoking Status']\n\nsns.barplot(x=['PC_1','PC_2'],y=pca.explained_variance_ratio_)\nsns.pointplot(x=['PC_1','PC_2'],y=np.cumsum(pca.explained_variance_ratio_),lw=5,legend=True,label='Cumulative',color='tab:red')\nplt.ylabel('Explained Variance')\nplt.title('Explained Variance Ratio After Projecting $R^{20} \\longrightarrow R^{2}$')\nplt.show()","metadata":{"execution":{"iopub.status.busy":"2021-11-10T05:29:25.252560Z","iopub.execute_input":"2021-11-10T05:29:25.252908Z","iopub.status.idle":"2021-11-10T05:29:25.908611Z","shell.execute_reply.started":"2021-11-10T05:29:25.252867Z","shell.execute_reply":"2021-11-10T05:29:25.907652Z"},"trusted":true},"execution_count":14,"outputs":[]},{"cell_type":"markdown","source":"**Observation:** Looking at the amount of variance we are able to preserve even after projecting our data to a 2-dimensional space confirms our prior hypothesis, being able to preserve more than 80% of the variance with just 2 Principal Components is amazing.","metadata":{}},{"cell_type":"code","source":"ax1 = plt.subplot(221)\nax1.set_title('$R^{2}$ Reduced Dimension of Our Genomic Representation of Each Sample',fontsize=15)\nsns.scatterplot(x=t_df['pc1'],y=t_df['pc2'],hue=t_df['Gender'],ax=ax1)\nax1.set_title('$R^{2}$ Reduced Dimension of Our Genomic Representation of Each Sample',fontsize=15)\nax2 = plt.subplot(223)\nsns.scatterplot(x=t_df['pc1'],y=t_df['pc2'],hue=t_df['Smoking Status'],ax=ax2)\nax3 = plt.subplot(122)\nsns.scatterplot(x=t_df['pc1'],y=t_df['pc2'],size=t_df['Age'],hue=t_df['Age'],ax=ax3)\nax2.set_title('$R^{2}$ Reduced Dimension of Our Genomic Representation of Each Sample',fontsize=15)\nax3.set_title('$R^{2}$ Reduced Dimension of Our Genomic Representation of Each Sample',fontsize=15)\n\nplt.tight_layout()\n\nplt.show()","metadata":{"execution":{"iopub.status.busy":"2021-11-10T05:29:44.787332Z","iopub.execute_input":"2021-11-10T05:29:44.788056Z","iopub.status.idle":"2021-11-10T05:29:46.129273Z","shell.execute_reply.started":"2021-11-10T05:29:44.788019Z","shell.execute_reply":"2021-11-10T05:29:46.127498Z"},"trusted":true},"execution_count":15,"outputs":[]},{"cell_type":"markdown","source":"**Observation:** now that we represent all our 20 probes with linear combinations consisting of 2 coefficients, we revisit the analysis we performed based on ridge plots in an earlier section, and here the massive difference between the genders is clear as day light!","metadata":{}},{"cell_type":"code","source":"plt.figure(figsize=(15,8))\nplt.title('Gender vs PC 1 Value')\nsns.scatterplot(x=t_df['pc1'],y=t_df['Gender'])\nplt.show()","metadata":{"execution":{"iopub.status.busy":"2021-11-10T05:29:58.532264Z","iopub.execute_input":"2021-11-10T05:29:58.532527Z","iopub.status.idle":"2021-11-10T05:29:58.811648Z","shell.execute_reply.started":"2021-11-10T05:29:58.532497Z","shell.execute_reply":"2021-11-10T05:29:58.810977Z"},"trusted":true},"execution_count":16,"outputs":[]},{"cell_type":"code","source":"plt.title('Different Values of Beta Example')\nx = np.linspace(-4, 4, 100)\nplt.plot(x, 1.0 / (1.0 + np.exp(-2 * x)), label=r\"$\\beta = -2$\",lw=3)\nplt.plot(x, 1.0 / (1.0 + np.exp(52 * x)), label=r\"$\\beta = 52$\",lw=3)\nplt.plot(x, 1.0 / (1.0 + np.exp(7 * x) ), label=r\"$\\beta = 7$\",lw=3)\nplt.legend();","metadata":{"execution":{"iopub.status.busy":"2021-11-10T05:30:18.047977Z","iopub.execute_input":"2021-11-10T05:30:18.048680Z","iopub.status.idle":"2021-11-10T05:30:18.424723Z","shell.execute_reply.started":"2021-11-10T05:30:18.048640Z","shell.execute_reply":"2021-11-10T05:30:18.424040Z"},"trusted":true},"execution_count":17,"outputs":[]},{"cell_type":"code","source":"with pm.Model() as model:\n    beta = pm.Normal(\"beta\", mu=0, tau=0.001, testval=0)\n    alpha = pm.Normal(\"alpha\", mu=0, tau=1/t_df.pc1.std(), testval=0)\n    p = pm.Deterministic(\"p_parm\", 1.0/(1. + tt.exp(beta*t_df.pc1 + alpha)))","metadata":{"execution":{"iopub.status.busy":"2021-11-10T05:30:33.127633Z","iopub.execute_input":"2021-11-10T05:30:33.128350Z","iopub.status.idle":"2021-11-10T05:30:47.236901Z","shell.execute_reply.started":"2021-11-10T05:30:33.128316Z","shell.execute_reply":"2021-11-10T05:30:47.235996Z"},"trusted":true},"execution_count":18,"outputs":[]},{"cell_type":"markdown","source":"Notice in the above code we had to set the values of beta and alpha to 0. The reason for this is that if beta and alpha are very large, they make p equal to 1 or 0. Unfortunately, pm.Bernoulli does not like probabilities of exactly 0 or 1, though they are mathematically well-defined probabilities. So by setting the coefficient values to 0, we set the variable p to be a reasonable starting value.","metadata":{}},{"cell_type":"code","source":"with model:\n    observed = pm.Bernoulli(\"obs\", p, observed=t_df.Gender)\n    start = pm.find_MAP()\n    step = pm.Metropolis()\n    trace = pm.sample(120000, step=step, start=start)\n    burned_trace = trace[100000::2]","metadata":{"execution":{"iopub.status.busy":"2021-11-10T05:30:50.681721Z","iopub.execute_input":"2021-11-10T05:30:50.682375Z","iopub.status.idle":"2021-11-10T05:33:44.016550Z","shell.execute_reply.started":"2021-11-10T05:30:50.682333Z","shell.execute_reply":"2021-11-10T05:33:44.015796Z"},"trusted":true},"execution_count":19,"outputs":[]},{"cell_type":"code","source":"alpha_samples = burned_trace[\"alpha\"][:, None]\nbeta_samples = burned_trace[\"beta\"][:, None]\nplt.subplot(211)\nplt.title(r\"Posterior distributions of the variables $\\alpha, \\beta$\")\nsns.histplot(beta_samples, bins=35, alpha=0.85,label=r\"posterior of $\\beta$\", palette=[\"#7A68A6\"],stat='probability')\nplt.legend()\n\nplt.subplot(212)\nsns.histplot(alpha_samples, bins=35, alpha=0.85,label=r\"posterior of $\\alpha$\", palette=[\"#A60628\"],stat='probability')\nplt.legend();","metadata":{"execution":{"iopub.status.busy":"2021-11-10T05:34:08.786577Z","iopub.execute_input":"2021-11-10T05:34:08.786870Z","iopub.status.idle":"2021-11-10T05:34:09.523090Z","shell.execute_reply.started":"2021-11-10T05:34:08.786831Z","shell.execute_reply":"2021-11-10T05:34:09.522387Z"},"trusted":true},"execution_count":20,"outputs":[]},{"cell_type":"code","source":"t = np.linspace(t_df.pc1.min() - 2, t_df.pc1.max()+2, 50)[:, None]\ndef logistic(x, beta, alpha=0):\n    return 1.0 / (1.0 + np.exp(np.dot(beta, x) + alpha))\np_t = logistic(t.T, beta_samples, alpha_samples)\n\nmean_prob_t = p_t.mean(axis=0)","metadata":{"execution":{"iopub.status.busy":"2021-11-10T05:34:33.301537Z","iopub.execute_input":"2021-11-10T05:34:33.301833Z","iopub.status.idle":"2021-11-10T05:34:33.319778Z","shell.execute_reply.started":"2021-11-10T05:34:33.301769Z","shell.execute_reply":"2021-11-10T05:34:33.318777Z"},"trusted":true},"execution_count":21,"outputs":[]},{"cell_type":"code","source":"plt.plot(t, mean_prob_t, lw=3, label=\"average posterior \\nprobability \\ of defect\")\nplt.plot(t, p_t[0, :], ls=\"--\", label=\"realization from posterior\")\nplt.plot(t, p_t[-2, :], ls=\"--\", label=\"realization from posterior\")\nplt.scatter(t_df.pc1, t_df.Gender, color=\"tab:red\", s=50, alpha=0.5)\nplt.title(\"Posterior expected value of probability of being Male; \\\nplus realizations\")\nplt.legend()\nplt.ylim(-0.1, 1.1)\nplt.xlim(t.min(), t.max())\nplt.ylabel(\"probability\")\nplt.xlabel(\"temperature\");","metadata":{"execution":{"iopub.status.busy":"2021-11-10T05:34:57.641555Z","iopub.execute_input":"2021-11-10T05:34:57.642241Z","iopub.status.idle":"2021-11-10T05:34:58.017286Z","shell.execute_reply.started":"2021-11-10T05:34:57.642201Z","shell.execute_reply":"2021-11-10T05:34:58.016628Z"},"trusted":true},"execution_count":22,"outputs":[]},{"cell_type":"code","source":"qs = mquantiles(p_t, [0.025, 0.975], axis=0)\nplt.fill_between(t[:, 0], *qs, alpha=0.7,color=\"#7A68A6\")\nplt.plot(t[:, 0], qs[0], label=\"95% CI\", color=\"#7A68A6\", alpha=0.7)\nplt.plot(t, mean_prob_t, lw=1, ls=\"--\", color=\"k\",\n         label=\"average posterior \\nprobability of defect\")\nplt.xlim(t.min(), t.max())\nplt.ylim(-0.02, 1.02)\nplt.legend()\nsns.scatterplot(x=t_df.pc1,y= t_df.Gender, color=\"tab:red\", s=50, alpha=0.5)\nplt.xlabel(\"$PC_1$, $X$\")\nplt.ylabel(\"probability estimate\")\nplt.title(\"Posterior probability estimates given $PC_1$ Value. $X$\");","metadata":{"execution":{"iopub.status.busy":"2021-11-10T05:35:31.172304Z","iopub.execute_input":"2021-11-10T05:35:31.172570Z","iopub.status.idle":"2021-11-10T05:35:31.613268Z","shell.execute_reply.started":"2021-11-10T05:35:31.172539Z","shell.execute_reply":"2021-11-10T05:35:31.612644Z"},"trusted":true},"execution_count":23,"outputs":[]},{"cell_type":"markdown","source":"The 95% credible interval, or 95% CI, painted in purple, represents the interval, for each  PC1  value, that contains 95% of the distribution. For example, at 0.01 , we can be 95% sure that the probability of being male between 0.98 and 0.99.The 95% credible interval, or 95% CI, painted in purple, represents the interval, for each  PC1  value, that contains 95% of the distribution. For example, at 0.01 , we can be 95% sure that the probability of being male between 0.98 and 0.99.","metadata":{}}]}